client/src/components/PaymentModal.jsx

```diff
@@ -0,0 +1,236 @@
+import { useEffect, useState } from "react";
+import Select from "react-select";
+import { useUserContext } from "../contexts/UserContextProvider";
+import { getUserCard } from "../utilities/cardApi";
+import {
+  generateCardExpiryYears,
+  generateMonths,
+} from "../utilities/helperFunctions";
+import { customStyles } from "../utilities/selectStyle";
+import { GrFormClose } from "react-icons/gr";
+import { buyProduct } from "../utilities/productsApi";
+import LoadingSpinner from "./LoadingSpinner";
+import { toast } from "react-toastify";
+import "react-toastify/dist/ReactToastify.css";
+
+const PaymentModal = ({ setShowModal, product, setProduct }) => {
+  const [cardDetails, setCardDetails] = useState({
+    cardHolderName: "",
+    cardNumber: "",
+    expirationYear: "",
+    expirationMonth: "",
+    cvc: "",
+  });
+  const [isLoading, setIsLoading] = useState(false);
+  const { user } = useUserContext();
+
+  useEffect(() => {
+    (async () => {
+      const response = await getUserCard(user.id);
+      setCardDetails({
+        cardHolderName:
+          response.data.cardHolderName === null
+            ? ""
+            : response.data.cardHolderName,
+        cardNumber:
+          response.data.cardNumber === null ? "" : response.data.cardNumber,
+      });
+    })();
+  }, []);
```

Comment:
This same use effect is repeated 3 times in different components, couldn't we fetch it together with user info, and keep in user context?